Add Sinkable and Channelable types. Fix Subscribable types.

vaikas-google opened the pull request:
Add Sinkable / Channelable types, clean up the Subscribable types.

mattmoor commented on apis/duck/v1alpha1/sinkable_types.go:
```diff
@@ -0,0 +1,104 @@
+	if err := json.Unmarshal(raw, r); err != nil {
+		return nil, err
+	}
+	return r, nil
```
I think we'll significantly reduce boilerplate by introducing:

```go
package duck

func FromUnstructured(obj *unstructured.Unstructured, target interface{}) error {
// reserialize obj into target
}
```

mattmoor replied:
In fact, with named return values, I think this can become a single line.

vaikas-google replied:
yep, agree, good idea, but I don't see how to make this into a single line however. but I'll play with it some.

mattmoor commented on apis/duck/v1alpha1/sinkable_types.go:
```diff
@@ -0,0 +1,104 @@
+
+// Sinkable is the schema for the sinkable portion of the payload
+type Sinkable struct {
+	DomainInternal string `json:"domainInternal,omitempty"`
```
How is Sinkable different from Targetable?

n3wscott replied:
We have an idea that they are different in the data-plane. But they look the same in the control plane.

vaikas-google replied:
Look at Line 17-25 above :) If it's not clear, lmk and I'll update. IIRC, I also moved it down here if it's easier to find, mayhaps I put it into a nonstd location above the package.

vaikas-google replied:
Also replicated the comment to targetable for extra visiblity.

mattmoor commented on apis/duck/v1alpha1/sinkable_types.go:
```diff
@@ -0,0 +1,104 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Status Sinkable `json:"status"`
```
I understand that you have Targetable as Status for legacy reasons, but you also indicate:
> // For new resources that are built from scratch, they should probably
> // introduce an extra level of indirection.

Maybe Sinkable also suffers the legacy burden of wanting to support Route/Service?  (see comment above about how this is different from Targetable)

Given that Eventing hasn't yet released, compatibility with knative/serving 0.1 is (likely?) not a requirement and we could modify knative/serving at HEAD to move the targetable stuff you need around.  We couldn't drop `status.domainInternal` (yet), but we could add a status field for `targetable`, `sinkable`, as needed.

I think my preference would be to be more prescriptive about `.status.targetable` and (quickly) move `knative/serving` at HEAD to implement that instead of living with legacy baggage forever.  WDYT?

vaikas-google replied:
Yeah, I send you that observation last night offline as well. I also made a comment to that effect on Sinkable about the extra level of indirection.
I think we should do it. However, I'd propose to change that in a subsequent PR so that I could make the change in Serving first, then backport this code to support that?

mattmoor commented on apis/duck/v1alpha1/subscribable_types.go:
```diff
@@ -17,65 +17,105 @@ limitations under the License.
 )
 
+const (
+	StatusKeyName = "subscribable"
```
Or be deleted :)

mattmoor commented on apis/duck/v1alpha1/channelable_types.go:
```diff
@@ -0,0 +1,96 @@
+// and if there are calls, or results follwoing on. You subsribe to me
+// and I give you events and you deal with them??
+// of Call or Result must be present.
+type ChannelSubscriberSpec struct {
```
You do this for `Subscribable -> Channelable` below, so `Channelable -> Sinkable` should probably follow the same pattern?

vaikas-google replied:
No, because I do not want Channels to have to worry about resolving. My thinking was that Channel just gets an endpoint and it does not do any resolving. So this would be domainInternal (or something like that) which gets put in here. Or maybe I misunderstand what you mean here.

mattmoor commented on apis/duck/v1alpha1/subscribable_types.go:
```diff
@@ -17,65 +17,105 @@ limitations under the License.
 // typically use this type to deserialize Subscribable ObjectReferences and
 // access the Subscribable data.  This is not a real resource.
-type Topic struct {
+type ChannelableRef struct {
```
The naming is confusing me without a greater understanding of the object model.

vaikas-google replied:
yeah, original thinking was that having it called would be confusing. It's not, changed this to Subscription to mirror others.

imikushin commented on apis/duck/v1alpha1/channelable_types.go:
```diff
@@ -0,0 +1,96 @@
+type Channelable struct {
+	// TODO: What is actually required here for Channel spec.
+	// This is the list of subscriptions for this channel.
+	Subscribers []ChannelSubsciberSpec `json:"subscribers,omitempty"`
```
Did you mean `ChannelSubscriberSpec`? (missing `r`)

imikushin commented on apis/duck/v1alpha1/channelable_types.go:
```diff
@@ -0,0 +1,96 @@
+func (t *Channel) Populate() {
+	t.Spec.Channelable = &Channelable{
+		// Populate ALL fields
+		Subscriptions: []string{"subscription1", "subscription2"},
```
Did you mean to set the field `Subscribers`?

vaikas-google replied:
yeppers, thanks!

vaikas-google commented:
Thanks for the reviews. I believe I addressed all the issues, PTAL.

 I can't figure out how to construct a failing marshaling of the unstructured, and seems beyond what's necessary for this.

mattmoor commented on apis/duck/v1alpha1/util.go:
```diff
@@ -0,0 +1,34 @@
+
+// FromUnstructured tkaes unstructured object from (say from client-go/dynamic) and
+// converts it into our duck types.
+func FromUnstructured(obj unstructured.Unstructured, target interface{}) error {
```
Oops, I meant to push this up a layer in the directory structure, since I don't see a strong reason to version this.

vaikas-google replied:
I don't think I can just move it there, because I need the types from v1alpha1 directory which imports ../ and then the test would fail. I moved it to apis/duck/util

mattmoor commented:
Also discussed offline adding parallel `LegacyTargetable` (what we have) and `Targetable` (what we want) in terms of using the nested structure pattern for things in status.

The former unblocks eventing from using serving `0.1.x`, and the latter sets things up so that we can implement `Targetable` in `knative/serving` at HEAD for `knative/eventing` to switch to.

vaikas-google commented on apis/duck/v1alpha1/channelable_types.go:
```diff
@@ -0,0 +1,94 @@
+// One of them must be present
+type ChannelSubscriberSpec struct {
+	CallableDomain string `json:"callableDomain"`
+	SinkableDomain string `json:"sinkableDomain"`
```
done

vaikas-google commented on apis/duck/util/util_test.go:
```diff
@@ -0,0 +1,104 @@
+
+	"encoding/json"
+
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
```
I'm not sure I follow. I couldn't put it in ../ because of go circular dependency import problem. This depends on the v1alpha1 types and it depends on ../

vaikas-google commented on apis/duck/util/util.go:
```diff
@@ -0,0 +1,34 @@
+limitations under the License.
+*/
+
+package util
```
I probably could've but seems like the ../ already defines it and I'm sure it would at some point lead to some hilarity. If you want, I can do that though :)

vaikas-google commented on apis/duck/util/util.go:
```diff
@@ -0,0 +1,34 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// FromUnstructured tkaes unstructured object from (say from client-go/dynamic) and
```
done

vaikas-google commented on apis/duck/v1alpha1/targetable_types.go:
```diff
@@ -45,12 +48,12 @@ type Target struct {
+// TargetableStatus shows how we expect folks to embed Targetable in
 // their Status field.
-type TargetStatus struct {
+type TargetableStatus struct {
```
done

vaikas-google commented on apis/duck/v1alpha1/subscribable_types.go:
```diff
@@ -17,65 +17,79 @@ limitations under the License.
+	// Channelable is a reference to the actual Channelable
+	// interface that provides the Subscription capabilities. This may
+	// point to object itself or to another object providing the
+	// actual implementation.
```
Thanks for the pointer, better?

vaikas-google commented on apis/duck/v1alpha1/subscribable_types.go:
```diff
@@ -17,65 +17,79 @@ limitations under the License.
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// Topic is a skeleton type wrapping Subscribable in the manner we expect
+// Subscribable is a skeleton type wrapping Subscribable in the manner we expect
```
Publishable seems not better to me either, so not going to churn on these names in code. Point is that there is an object that you can subscribe to to receive events from, it should be subscribable imho. I tried to clarify in the comments. lmk if that doesn't make it better.

mattmoor commented:
/lgtm
